cipher/src/stream_wrapper.rs

```diff
@@ -0,0 +1,231 @@
+                // double slicing allows to remove panic branches
+                data.xor_in2out(&self.buffer[pos..][..n]);
+                self.set_pos_unchecked(pos + n);
+                return Ok(());
+            }
+            let (mut left, right) = data.split_at(r);
+            data = right;
+            left.xor_in2out(&self.buffer[pos..]);
+        }
+
+        let (blocks, mut leftover) = data.into_chunks();
+        self.core.apply_keystream_blocks_inout(blocks);
+
+        let n = leftover.len();
+        if n != 0 {
+            self.core.write_keystream_block(&mut self.buffer);
+            leftover.xor_in2out(&self.buffer[..n]);
+        }
+        self.set_pos_unchecked(n);
+
+        Ok(())
+    }
+}
+
+impl<T: StreamCipherSeekCore> StreamCipherSeek for StreamCipherCoreWrapper<T>
+where
+    T::BlockSize: IsLess<U256>,
+    Le<T::BlockSize, U256>: NonZero,
+{
+    fn try_current_pos<SN: SeekNum>(&self) -> Result<SN, OverflowError> {
+        let Self { core, pos, .. } = self;
+        SN::from_block_byte(core.get_block_pos(), *pos, T::BlockSize::U8)
+    }
+
+    fn try_seek<SN: SeekNum>(&mut self, new_pos: SN) -> Result<(), StreamCipherError> {
+        let Self { core, buffer, pos } = self;
+        let (block_pos, byte_pos) = new_pos.into_block_byte(T::BlockSize::U8)?;
+        core.set_block_pos(block_pos);
+        if byte_pos != 0 {
+            self.core.write_keystream_block(buffer);
+        }
+        *pos = byte_pos;
+        Ok(())
+    }
+}
+
+// Note: ideally we would only implement the InitInner trait and everything
+// else would be handled by blanket impls, but unfortunately it will
+// not work properly without mutually exclusive traits, see:
+// https://github.com/rust-lang/rfcs/issues/1053
+
+impl<T: KeySizeUser + BlockSizeUser> KeySizeUser for StreamCipherCoreWrapper<T>
+where
+    T::BlockSize: IsLess<U256>,
+    Le<T::BlockSize, U256>: NonZero,
+{
+    type KeySize = T::KeySize;
+}
+
+impl<T: IvSizeUser + BlockSizeUser> IvSizeUser for StreamCipherCoreWrapper<T>
+where
+    T::BlockSize: IsLess<U256>,
+    Le<T::BlockSize, U256>: NonZero,
+{
+    type IvSize = T::IvSize;
+}
+
+impl<T: KeyIvInit + BlockSizeUser> KeyIvInit for StreamCipherCoreWrapper<T>
+where
+    T::BlockSize: IsLess<U256>,
+    Le<T::BlockSize, U256>: NonZero,
+{
+    #[inline]
+    fn new(key: &Key<Self>, iv: &Iv<Self>) -> Self {
+        Self {
+            core: T::new(key, iv),
+            buffer: Default::default(),
+            pos: 0,
+        }
+    }
+}
+
+impl<T: KeyInit + BlockSizeUser> KeyInit for StreamCipherCoreWrapper<T>
+where
+    T::BlockSize: IsLess<U256>,
+    Le<T::BlockSize, U256>: NonZero,
+{
+    #[inline]
+    fn new(key: &Key<Self>) -> Self {
+        Self {
+            core: T::new(key),
+            buffer: Default::default(),
+            pos: 0,
+        }
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl<T> Zeroize for StreamCipherCoreWrapper<T>
+where
+    T: BlockSizeUser + Zeroize,
```

Comment:
@tarcieri 
IIUC we can not implement `ZeroizeOnDrop` for this wrapper, so users would have to rely on manually wrapping it with `Zeroizing`, correct?

Reply:
You could potentially impl a `Drop` handler that calls `zeroize` on the fields, as well as manually impl `ZeroizeOnDrop` as a marker trait, but it would be feature-gated, which would be a little weird.

Impling `ZeroizeOnDrop` would require bumping the `zeroize` version to 1.5.

Reply:
No, the issue is that `Drop` impl requires the same bounds as struct definition. We would have to bind the wrapper on `T: Zeroize` with enabled `zeroize` feature, meaning that build could break if it's used with a type which does not implement `Zeroize`.

Reply:
Aah yes, the bounds need to be consistent between the `Drop` impl and the struct

Reply:
I've decided to implement zeroizing `Drop`, but not `Zeroize` itself. It looks like we should not implement both of them simultaneously. Having `Zeroize` implemented on complex types (e.g. cipher states) can result in users erroneously zeroizing them, which can lead to security vulnerabilities.

The `Drop` impl zeroizes only buffer and cursor, while zeroization of the core type `T` is left for its own `Drop` impl. Implementation of `ZeroizeOnDrop` is gated on `T: ZeroizeOnDrop`.